pkg/authz/responsefilterer.go

```
@@ -0,0 +1,404 @@
+		case 1:
+			if !strings.HasSuffix(gvk.Kind, "List") {
+				gvk.Kind = gvk.Kind + "List"
+			}
+
+			var typedList runtime.Object
+
+			if recognized {
+				var err error
+				typedList, err = scheme.Scheme.New(gvk)
+				if err != nil {
+					return fmt.Errorf("failed to create new object of type %s: %w", gvk, err)
+				}
+			} else {
+				// custom types
+				typedList = &unstructured.UnstructuredList{}
+			}
+			_, _, err = decoder.Decode(body, &gvk, typedList)
+			if err != nil {
+				return fmt.Errorf("failed to decode response body: %w", err)
+			}
+			filterErr = rf.filterList(typedList, result)
+			if filterErr != nil {
+				break
+			}
+			if recognized {
+				filterErr = encoder.Encode(typedList, &filteredBody)
+			} else {
+				// custom types
+				filterErr = unstructured.UnstructuredJSONScheme.Encode(typedList, &filteredBody)
+			}
+			if filterErr != nil {
+				break
+			}
+
+		// if there's 2 or parts in the url (i.e. "pods/foo", "pods/foo/status"), we assume it's a single object
+		default:
+			var typedObj runtime.Object
+
+			if recognized {
+				var err error
+				typedObj, err = scheme.Scheme.New(gvk)
+				if err != nil {
+					return fmt.Errorf("failed to create new object of type %s: %w", gvk, err)
+				}
+			} else {
+				// custom types
+				typedObj = &unstructured.Unstructured{}
+			}
+			_, _, err = decoder.Decode(body, &gvk, typedObj)
+			if err != nil {
+				return fmt.Errorf("failed to decode response body: %w", err)
+			}
+
+			filterErr = rf.filterObject(typedObj, result)
+			if filterErr != nil {
+				break
+			}
+			filteredBody = *bytes.NewBuffer(body)
+		}
+
+		return writeResp(filteredBody, filterErr, resp)
+	}
+}
+
+func writeResp(filteredBody bytes.Buffer, filterErr error, resp *http.Response) error {
+	// if there was an error, replace the body with an error message
+	if filterErr != nil {
+		filteredBody.Reset()
+		var merr error
+		errBody, merr := kjson.Marshal(k8serrors.NewUnauthorized(filterErr.Error()))
```

I understand this predates the PR. I think this makes transient errors difficult to troubleshoot. I'd rather return a 500 error, as I fear this may get confused with normal operation of the proxy.

---

Yeah, I didn't want to change this behavior